core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -778,4 +768,37 @@ private void failStuckSplitTasks()
             }
         }
     }
+
+    public static final class SqlTaskWithResults
+    {
+        private final SqlTask task;
+        private final ListenableFuture<BufferResult> resultsFuture;
+
+        public SqlTaskWithResults(SqlTask task, ListenableFuture<BufferResult> resultsFuture)
+        {
+            this.task = requireNonNull(task, "task is null");
+            this.resultsFuture = requireNonNull(resultsFuture, "resultsFuture is null");
+        }
+
+        public void recordHeartbeat()
+        {
+            task.recordHeartbeat();
+        }
+
+        public String getTaskInstanceId()
+        {
+            return task.getTaskInstanceId();
+        }
+
+        public boolean isTaskFailed()
+        {
+            TaskState state = task.getTaskState();
+            return (state == TaskState.ABORTED | state == TaskState.FAILED);
```

Comment:
```suggestion
            return (state == TaskState.ABORTED || state == TaskState.FAILED);
```

Reply:
Expanded to a `switch` instead since future work will consider more states as "failed".